### ISSUE TITLE
.NET 7 application hangs during GC

### ISSUE THREAD

**Issue by KenKozman-GE** [dotnet/runtime#80073]
### Description

We host our application in Azure App Service in the host dependent mode.

Each instance of our application can end up getting stuck in a busy loop where it eats 100% of the CPU and no other HTTP calls are able to make it into the app. This only started happening when we upgraded to .NET 7.

It looks like this is due to one thread being stuck performing a GC, so all the other managed threads are paused.

Our application has health checks that run against it. One of those checks runs ~1/second. This check ends up querying a pgsql database and pulling down records (10s-1000s). This is the place we have seen the hang occur, although this same code can be called by other parts of the system. Likely it is just the health check that is triggering due to the much higher frequency of that call being made.

Here are a few of the unmanaged portion of the GC allocate from two separate dumps taken during the high CPU times (this is me doing dump debugging in Visual Studio 2022):

![image](https://user-images.githubusercontent.com/26578875/210100743-a5bcee5c-f177-422f-85a0-4505089d1f1b.png)
![image](https://user-images.githubusercontent.com/26578875/210100765-e4bfcf2a-15a0-44f0-b68e-a58b65428d27.png)

I believe it is running the Workstation GC based on the `WKS::` namespace (and because the other threads are frozen waiting on the GC and because the GC is being performed directly on the user thread). I believe this is happening because the app service plan instance shows up as having 1 processor. We specify to use the Server GC in our configuration but I believe it will use the workstation one if there is only one processor detected.

In one case I generated three dumps 5 minutes apart. In all three cases the managed code had not moved and was in the same exact place. The second unmanaged stack above is the top of the stack in those cases (although the very top most function changes in each, which is why I thought allocate_in_condemned_generations was the culprit). Here is the managed portion of the stack from those dumps:

![image](https://user-images.githubusercontent.com/26578875/210101074-e76266f9-5f60-4ef9-a1b6-ea2fb41a544d.png)

Here are some outputs from the CrashHangAnalysis MHT which AppService also kindly created (this is from one of the three dumps mentioned):

![image](https://user-images.githubusercontent.com/26578875/210101386-7ac80a35-36a3-4cf1-b64d-b54d7c483789.png)

As you can see in the thread times almost all of the time was eaten up by Thread 48. In each successive 5 minute dump you can see it has eaten up 5 more minutes of CPU time:

![image](https://user-images.githubusercontent.com/26578875/210101547-daa1e524-ed13-468f-8763-a91d5f54dedf.png)

Here is the full call stack for Thread 48 reported by the CrashHangAnalysis:

![image](https://user-images.githubusercontent.com/26578875/210101629-4b7935f5-acc0-490d-aef0-7a2fb52a347d.png)

There are also multiple CLR runtimes loaded in the process. I believe that AppService itself loads the .NET Framework 4.8, we are using .NET 7.0.0:

![image](https://user-images.githubusercontent.com/26578875/210101473-5956e6b5-b18b-4e0a-94e4-9ec878affa44.png)

I attempted to find this issue reported already but my searching turned up nothing :( Apologies if I missed it.

### Reproduction Steps

At this point I don't have a way to reproduce it reliably. One of our systems seems to be falling into this particular hole once a day or so. The SQL query on that system is a bit larger than other systems, but not massively so.

### Expected behavior

To just have the GC complete its work and return to our application.

### Actual behavior

It appears the GC gets stuck in WKS::gc_heap::allocate_in_condemned_generations. I see a retry label in there, perhaps it is stuck in an infinite loop?

### Regression?

I believe so. We never saw this on .NET 6 (several sub-versions) and .NET 5. We ran similar code with earlier versions of .NET Core and never saw any issues either.

### Known Workarounds

Perhaps downgrading to .NET 6?
It is uncertain if this would happen with .NET 7.0.1, I have not gotten to test that yet.
It is unclear if this is a workstation GC issue only; my guess is yes.

### Configuration

.NET 7.0.0 (in the dumps this shows up as CLR version v7.0.22).
Azure App Service Platform version: 99.0.7.620
Windows Server 2016 - 14393
Number of Processors - 1
Architecture: x64

### Other information

It *seems* like some sort of infinite loop issue in the `gc_heap::allocate_in_condemned_generations` function. But that is me using my "Jump to Conclusions" mat.

**Comment by cshung** [comment]
@KenKozman-GE Can you share a dump or something useful so that we can investigate?

**Comment by KenKozman-GE** [comment]
Yeah sorry...

Here are the crash hung analyses:
[CrashHangAnalyses.zip](https://github.com/dotnet/runtime/files/10326050/CrashHangAnalyses.zip)

Let me see if I can push the dumps here or not (they are full memory dumps, so quite large).

**Comment by KenKozman-GE** [comment]
Ah yep, 25MB limit. Let me try putting them somewhere that I can share, one second.

**Comment by KenKozman-GE** [comment]
Okay here are a couple of full memory dumps (mentioned above) as well as the crash hang analyses.
https://drive.google.com/drive/folders/1bcEIaf9KnGjlP_Qg1Qvo2AiPwy_py-Sf?usp=sharing
https://drive.google.com/file/d/1TVh_foyhMx4S_2shbpPeFz8Y_ab1sbnj/view?usp=sharing
https://drive.google.com/file/d/1yUKHTrQF2dXVfmVupatyI7MMg1teBxG8/view?usp=sharing

**Comment by KenKozman-GE** [comment]
One other workaround here would be to switch to an App Service Plan SKU which would show up as having multiple processors and thus use the Server GC (which would hopefully not exhibit the same issue).

Right now we are using P1V2 for some instances which have a single vCPU.

**Comment by KenKozman-GE** [comment]
One further update, we used to use S2 SKUs and switched to P1V2 in the middle of 2022. So this still seems like a regression from .NET 6 to .NET 7 for us, for the Workstation GC. But we would have been using the Server GC at the start of 2022 when we were still on S2 SKUs. We were on S2 SKUs during the .NET 5 timeframe as well.

**Comment by cshung** [comment]
I took a quick look at the dump and hypothesize that it may be an infinite loop because we forgot about the possibility of pinning.

Roughly speaking, the `allocate_in_condemned_generations` method does something along this line:

```
allocate_in_condemned_generations(...size...)
    retry:
    if we hit a pin or run out of space
        if we hit a pin
            deal with it
        if we run out of space
            deal with it
        goto retry
    ...
```

The key idea is that the `deal with it` should deal with the situation so that we should not be running into a pin or running out of space, as such, the loop is meant to be run for at most 3 iterations or less. (1 for pin, 1 for space, last for succeed). 

There are no synchronization primitives involved in this loop, so this is pretty much a case of an infinite loop.

There are no more pins in the pin queue (i.e. `gc_heap::mark_stack_bos == gc_heap::mark_stack_tos` (c.f. `gc_heap::pinned_plug_que_empty_p`)), therefore, the problem has something to do with space.

Looking at the implementation of `gc_heap::size_fit_p`, I understand why we get stuck running out of space.

We have a just-fit allocation context:

```
0:046> ?? gen->allocation_context.alloc_ptr
unsigned char * 0x000001d2`72c00020

0:046> ?? gen->allocation_context.alloc_limit
unsigned char * 0x000001d2`73000000

0:046> ?? size
unsigned int64 0x3fffe0
```

Note that `0x000001d272c00020 + 0x3fffe0 == 0x000001d273000000`, the allocation context should fit the allocation exactly, probably prepared so by the `deal_with_it` code for space and that's why it is so exact.

But then `old_loc != nullptr`, so the code in `size_fit_p` insists on requiring a bit more memory (i.e. `+ Align(min_object_size)`), so that check will fail.

We are probably stuck in that situation and that is why the loop isn't breaking. What baffles me is that the `old_loc` is already considered in the `deal with it` section for running out of space by being passed as an argument to `grow_heap_segment`, not sure why we aren't processing that correctly.

**Comment by Maoni0** [comment]
> it may be an infinite loop because we forgot about the possibility of pinning.

I don't know how to interpret this. `allocate_in_condemned_generations` has to handle pinning; otherwise every .net app would fail.

from the debug info given above, it doesn't look like it has to do with pinning - it says we have a huge plug where every single object on that region survived. and we are asking for an additional min object size when fitting in this case. can you check to see if it's asking for an additional min_obj_size because of `USE_PADDING_FRONT` or `USE_PADDING_TAIL`? what puzzled me is it seems like it shouldn't be due to `USE_PADDING_FRONT` because the plug is so huge which tells me this plug should belong to gen2 so this should be a gen2 GC. but if it's a gen2 GC, we should not be passing `USE_PADDING_FRONT`. and we should not be passing `USE_PADDING_TAIL` because `generation_allocation_limit (gen)` should be the same as `heap_segment_plan_allocated (seg)`.

**Comment by cshung** [comment]
> can you check to see if it's asking for an additional min_obj_size because of USE_PADDING_FRONT or USE_PADDING_TAIL?

It is because of `USE_PADDING_FRONT`.

```
0:046> dv
       ...
       pad_in_front = 0n1
       ...
```

> what puzzled me is it seems like it shouldn't be due to USE_PADDING_FRONT because the plug is so huge which tells me this plug should belong to gen2 so this should be a gen2 GC.

It is a gen1 GC

```
0:046> ?? wks::gc_heap::settings
class WKS::gc_mechanisms
   +0x000 gc_index         : 0x6dc
   +0x008 condemned_generation : 0n1
   ...
```

This is on thread `46` of `RD501AC5AE7597_w3wp_5784_638079619256436212.dmp`

**Comment by Maoni0** [comment]
I looked at the 2 dumps and the really odd thing with both dumps is the first region in gen0 has no free objects whatsoever but all the other gen0 regions look normal, ie with free objects. this explains why we have such a huge plug in a gen1 GC but this is not what should happen - we are supposed to have free objects as padding between alloc contexts so we don't form potentially huge plugs like this.

@KenKozman-GE  do you happen to have a dump where it's running normally, ie, not getting this hang? that would be helpful.

**Comment by KenKozman-GE** [comment]
I can make one, let me go do that, one second.

**Comment by Maoni0** [comment]
thank you!

**Comment by KenKozman-GE** [comment]
Apologies the download and re-upload process here is somewhat onerous (IT security goons making sure I am not downloading or uploading various secrets and/or malware no doubt). Should be done in maybe an hour.

**Comment by KenKozman-GE** [comment]
Random possible workaround (if it is indeed a GC regression, we shall see): we could use the `clrgc.dll` as Maoni calls out [here](https://maoni0.medium.com/is-the-regression-in-gc-or-something-else-38f10018dd21).

I did not test that before filing the issue here, apologies for that.

**Comment by KenKozman-GE** [comment]
Okay here is an example dump of "normal processing": [dump](https://drive.google.com/file/d/1TGCibtw89qtEO5g4lVaN1pw6sMgKj6rb/view?usp=sharing)

**Comment by Maoni0** [comment]
yep, this one looks normal. if I look at the first region in gen0, I see 513 free objects (a region is 4mb and each alloc context is about 8k). other gen0 regions look fine too.

so something is causing us to not allocate these free objects in that 1st gen0 region in the bad case. via a brief code review I don't see how that can happen. I'll take another look tomorrow but I wanted to ask - would it be possible to use a private version of clrgc.dll to help with debugging this if it comes to that? you could of course trying use the shipped version of clrgc.dll by setting COMPlus_GCName=clrgc.dll (which would revert back to the .net 6.0 behavior) but we'd like to figure out why the 7.0 GC is failing for you. we could share a private version of clrgc.dll that includes some instrumentation to help and you can use via the same way. would that be feasible? thanks!

**Comment by KenKozman-GE** [comment]
Oh yeah we could do that I think.

I only mention the clrgc.dll bit above to help me remember and in case others see this type of issue in the future.

Anyway, just let me know and we can try to help however.

**Comment by Maoni0** [comment]
just an update - I'm doing some testing on my side for the private build and also trying to see if I can repro this. will let you know when it's ready for you to pick up.

**Comment by KenKozman-GE** [comment]
Sounds good. We have seen it happen still once every day or two on at least one tenant. So hopefully we can get a useful dump pretty quickly.

**Comment by Maoni0** [comment]
if I make it so that it incurs an access violation when it detects this situation, would your system capture a dump? or would it be easier if I also make it hang? I usually do an AV in this situation but wanted to check if that's convenient for you.

**Comment by KenKozman-GE** [comment]
I think an AV will be easiest. The hosting we use (Azure App Service) will capture a dump and restart it in that case.

This seems nicer than just making it hang given less potential downtime for the one instance.

**Comment by Maoni0** [comment]
cool! I have the bits ready at https://github.com/Maoni0/clrgc/tree/main/issues/80073. there's a readme that explains how to use the files (in the "How to test" section). I also included explanation of changes included in the private builds and src/symbols in case you want to look at/use them.

**Comment by KenKozman-GE** [comment]
Thanks @Maoni0! I likely won't be able to install it and test it until next Tuesday at the earliest (vacation, etc.). But will try to get to it as soon as I am able.

**Comment by Maoni0** [comment]
that's totally fine, thanks!

**Comment by KenKozman-GE** [comment]
Okay sorry @Maoni0 I am just getting back to this. I've never tried to install/use a `clrgc.dll`. I'm realizing now I might not be able to do this due to us deploying as a [Framework Dependent](https://learn.microsoft.com/en-us/dotnet/core/deploying/#publish-framework-dependent) DLL and dependencies to Azure App Service.

I say this as I see no coreclr.dll anywhere I can make changes. So I think it is just in the system area. Sounds like the `clrgc.dll ` has to be in the same dir as coreclr.dll?

Also I've noticed that some optimization work we deployed Monday has seemed to make it where we aren't getting those hangs anymore, likely due to just much less GC pressure.

I can add back the inefficient code to a QA system and try to test there, but I'm also not sure how exactly to go about getting our bits to use the bits you made. Sounds like the clrgc.dll *has* to be in the same dir as coreclr.dll?

I could try to publish a self-contained piece for this, but I have never done that.

**Comment by Maoni0** [comment]
yes clrgc.dll needs to be in the same dir as coreclr.dll. you are using coreclr.dll from `D:\Program Files\dotnet\shared\Microsoft.NETCore.App\7.0.0\coreclr.dll`. would it be possible for you to replace the clrgc.dll in that dir with the one I provided? this doesn't affect anyone unless they are also specifying to use clrgc.dll.

**Comment by KenKozman-GE** [comment]
Yeah sorry I noticed after I posted that this was the path specified in the dump.

I am "over my skis" talking about the App Service bit themselves. But it looks like I don't have permission to write files in that directory. It [sounds like](https://github.com/projectkudu/kudu/wiki/Understanding-the-Azure-App-Service-file-system#machine-level-read-only-files) access to these is read only. I imagine this is due to the sharing model of the underlying layers of infrastructure.

So I'm not sure if there is anything I can do with our current deployment method?

**Comment by Maoni0** [comment]
if you have admin privilege you can certainly replace this file (or any file in the shared dir). do you think you have admin privilege?

**Comment by KenKozman-GE** [comment]
I don't believe I do. This is vanilla Azure App Service and the docs seem to indicate all of this is read only :(

**Comment by gldraphael** [comment]
@KenKozman-GE since you're going to setup a QA environment to test this anyway it might be worth trying a self-contained deployment as opposed to a framework dependent deployment like you mentioned. If you're going to deploy to your QA-env through VS, VS takes care of everything for you, just pick self-contained at the time of deployment.

**Comment by KenKozman-GE** [comment]
Yeah that is what I was poking around at. We don't do direct VS deploys, but I think I should be able to make this work using our current methods.

I'll try this out today (assuming I can find time). Sorry for the delay here.

**Comment by Maoni0** [comment]
@KenKozman-GE please don't apologize - you are doing work to help us find the root cause! I appreciate that you are trying to set up a self contained app. I hope you can still repro with self contained. thank you so much!

**Comment by cshung** [comment]
@KenKozman-GE, as GC bugs go, the repro is usually very sensitive to any change. If you could make it as close as it was, that would optimize for the possibility of reproducing the issue.

**Comment by KenKozman-GE** [comment]
Yeah that makes sense. I will try to make that happen.

**Comment by KenKozman-GE** [comment]
Okay after much learning and failing on my part I believe I have an environment up and running with the `with_change` clrgc.dll.

Here is a screen shot of a dump from when it was behaving well:
![image](https://user-images.githubusercontent.com/26578875/212439339-da96e186-72fb-4738-827d-8a47ed52111c.png)

Now we just have to wait for it to fail :)

**Comment by KenKozman-GE** [comment]
Okay I have made additional changes based on @cshung's recommendation to try to return this to how it was when it was happening more reliably.

Still no reproductions, but hopefully we will catch one soon.

**Comment by KenKozman-GE** [comment]
Hrm... well so far no dice on having it replicate.

We have also seen it stop replicating on the two instances it was consistently happening on. We made some code changes which would have reduced a lot of memory pressure though, so I am not surprised it stopped happening there. I rolled back those changes for the test instance before trying to get it to fail... but so far no luck.

**Comment by KenKozman-GE** [comment]
Awww yeah, looks like we got a breakpoint "crash". I think this is hopefully what you guys need :)

Dump available [here](https://drive.google.com/file/d/1Hp8rif7QUmSTTKx0R4FCmad47n5uhlvm/view?usp=sharing).

**Comment by Maoni0** [comment]
thank you! I'm downloading the dump now and will take a look tonight.

**Comment by Maoni0** [comment]
sorry I didn't end up getting time to look at this last night. so I took a look just now. unfortunately you are hitting a different issue. @cshung the logic at the end of plan phase that we introduced in [#76674](https://github.com/dotnet/runtime/pull/76674) where we check for special_sweep_p in all heaps and set gc_policy accordingly if it's TRUE needs to be available for Workstation GC as well. in the dump you could see that special_sweep_p is TRUE, but we were still doing a compacting GC which causes gen0 to end up with 0 regions.

**Comment by KenKozman-GE** [comment]
Yikes okay. So I should keep our test running?

One other thought/question: is it possible to force the server GC even in a single core setup? That might be a nice "feature" for the user in cases like this. (I suggest this assuming in a single core the server GC is likely a de-optimization and/or that there are loads of other factors I have no idea about.)

**Comment by Maoni0** [comment]
> One other thought/question: is it possible to force the server GC even in a single core setup? 

you could set DOTNET_PROCESSOR_COUNT to 2 on a single core setup and set it to use Server GC and GCHeapCount to 1. but of course that means any code you use based on Environment.ProcessCount would be 2 so that may or may not cause problems. I wouldn't do that though because we don't test Server GC with 1 heap - it probably runs fine but it's simply an unsupported scenario. we should just fix this bug with Workstation GC.

**Comment by KenKozman-GE** [comment]
Ah, yeah I was thinking just some GC flag to force it to use Server vs. Workstation, even in a single core scenario.

But it sounds like that isn't a great idea as it is (as you mentioned) an unsupported scenario. To support that on y'alls end would probably not be worth it. I was hoping it might be a non-item for you all to support, but I am not surprised it isn't.

No worries :) We can always just switch our underlying SKU to be a multi-core one if the problem starts to be too painful for us (it hasn't cropped up in the production stacks since a few days after Christmas).

**Comment by Maoni0** [comment]
thank you so much for being so accommodating! @cshung is working on a fix for this as we speak.

**Comment by KenKozman-GE** [comment]
No worries! I am just glad we are able to help at all.

We all really dig .Net and appreciate all the great stuff you guys do.

**Comment by KenKozman-GE** [comment]
Okay when I checked in this morning I saw the process was chewing up the CPU again on our test-beta stack. It did not have an intentional/breakpoint "crash" though.

I also did not have CPU monitoring running (the monitor window had elapsed). I attempted to capture a manual dump but I'm not sure that worked. The .dmp file is [here](https://drive.google.com/file/d/1P7qzhUjig3TEK_jw6XF_Iav7ZXpRYXNU/view?usp=sharing) but I was unable to get VS to debug it in managed mode. I didn't try any fancier windmp or other methods. Maybe this one will be helpful though?

The process looks like it restarted right around when I was trying to dump it, so it could be this is not even the infinite loop process.

Let me know if this one helps.

**Comment by Maoni0** [comment]
thanks! this dump has no GC or allocation work going on at all. the top 5 threads that consumed the most CPU (mostly in user mode) are (the other threads spent a tiny amount, much less than 1s) -
```
 User Mode Time
  Thread       Time
   29:e24      0 days 0:00:04.250
   46:1dc4     0 days 0:00:03.218
    4:1cfc     0 days 0:00:02.468
   38:222c     0 days 0:00:02.359
   43:9cc      0 days 0:00:02.000
```
so even the top thread only had ~4s in CPU. they are currently in one of the 2 call stacks -

```
 # Child-SP          RetAddr               Call Site
00 0000000d`a16be848 00007ff8`0cf44cdf     ntdll!ZwWaitForMultipleObjects+0x14
01 0000000d`a16be850 00007fff`e132cd6c     KERNELBASE!WaitForMultipleObjectsEx+0xef
02 (Inline Function) --------`--------     coreclr!Thread::DoAppropriateAptStateWait+0x54 [D:\a\_work\1\s\src\coreclr\vm\threads.cpp @ 3380] 
03 0000000d`a16beb50 00007fff`e132cba9     coreclr!Thread::DoAppropriateWaitWorker+0x184 [D:\a\_work\1\s\src\coreclr\vm\threads.cpp @ 3514] 
04 0000000d`a16bec30 00007fff`e132abf8     coreclr!Thread::DoAppropriateWait+0x89 [D:\a\_work\1\s\src\coreclr\vm\threads.cpp @ 3229] 
05 (Inline Function) --------`--------     coreclr!CLREventBase::WaitEx+0x36 [D:\a\_work\1\s\src\coreclr\vm\synch.cpp @ 459] 
06 (Inline Function) --------`--------     coreclr!CLREventBase::Wait+0x41 [D:\a\_work\1\s\src\coreclr\vm\synch.cpp @ 412] 
07 (Inline Function) --------`--------     coreclr!Thread::Wait+0x41 [D:\a\_work\1\s\src\coreclr\vm\threads.cpp @ 3992] 
08 (Inline Function) --------`--------     coreclr!Thread::Block+0x49 [D:\a\_work\1\s\src\coreclr\vm\threads.cpp @ 3949] 
09 0000000d`a16becb0 00007fff`e132a9cd     coreclr!SyncBlock::Wait+0x1c8 [D:\a\_work\1\s\src\coreclr\vm\syncblk.cpp @ 2871] 
0a (Inline Function) --------`--------     coreclr!ObjHeader::Wait+0x42 [D:\a\_work\1\s\src\coreclr\vm\syncblk.cpp @ 2270] 
0b (Inline Function) --------`--------     coreclr!Object::Wait+0x42 [D:\a\_work\1\s\src\coreclr\vm\object.h @ 337] 
0c 0000000d`a16bedd0 00007fff`824a70a0     coreclr!ObjectNative::WaitTimeout+0xcd [D:\a\_work\1\s\src\coreclr\classlibnative\bcltype\objectnative.cpp @ 266] 

 # Child-SP          RetAddr               Call Site
00 0000000d`ac01f298 00007ff8`0cf5681f     ntdll!ZwRemoveIoCompletion+0x14
01 0000000d`ac01f2a0 00007fff`fd98f811     KERNELBASE!GetQueuedCompletionStatus+0x3f
02 0000000d`ac01f300 00007fff`8450296e     picohelper!SetSpecializationTokenInProcess+0x59a1
03 0000000d`ac01f7c0 00000000`00000000     0x00007fff`8450296e

```

**Comment by Simon-Gregory-LG** [comment]
Hi, I'd like to add in to this discussion, as I think we are also experiencing this (or something very similar). Firstly I'd like to say it's good to see the helpful and responsive interactions on this thread so far.

We upgraded one of our applications from .NET 5.0 => .NET 7.0 and we have instances become unresponsive on at least 4 occasions since that time in various AppServices - it takes about 4-36 hours to occur. 

We have a 2 instance P2V3 and a single instance P2V3 in different environments that experience the same issue. We know the App usually has generally high memory utilisation / allocations which we have yet to get round to optimising yet but we didn't have this issue in .NET 5.0. On the multi-instance one, the instances seem to hang independently until eventually they are all hung and the app becomes completely unresponsive.

You can see that the health checks simply stopped responding on this one (the single instance AppService) two days ago:

![image](https://user-images.githubusercontent.com/45365503/216041589-3072477f-4472-455e-88f5-7a8c3b2c1574.png)

So, following the extremely useful thread above, I've done my own crash dump and the MHT shows the following:

![image](https://user-images.githubusercontent.com/45365503/216041920-022336f0-2b79-4e3d-ac1d-39df9592d194.png)

The offending thread 27 is in the GC:

![image](https://user-images.githubusercontent.com/45365503/216042041-681b032b-6988-4fc2-be30-105f9c8327fa.png)


At the moment we seem to be able to reproduce this fairly frequently. I don't have much available time, but anything I can do to help, let me know. In the mean-time we will have to change our plans and maybe go with .NET 6.0 until we're sure this is resolved.

Let me know if you think this is the same issue - otherwise I will raise a separate ticket.

Simon

**Comment by KenKozman-GE** [comment]
Oh wow interesting (to me playing armchair quarterback). The fact that this is in the Server GC vs. the Workstation GC is not something we had seen. Although we have far fewer instances running with more than a single core.

**Comment by KenKozman-GE** [comment]
Oh wow interesting (to me playing armchair quarterback). The fact that this is in the Server GC vs. the Workstation GC is not something we had seen. Although we have far fewer instances running with more than a single core.

**Comment by Maoni0** [comment]
thanks @Simon-Gregory-LG for reporting this! would it be possible to share a couple of dumps so I can check that you are seeing the same problem that @KenKozman-GE was seeing (most likely but I wanted to make sure)? also, would it be feasible at all to use the clrgc.dll that I shared? this was described in [this comment](https://github.com/dotnet/runtime/issues/80073#issuecomment-1374387080) above and would require you to be able to replace the clrgc.dll you are using (and set an env var to use it).

**Comment by Simon-Gregory-LG** [comment]
Yes, I can try do that. Is there an easy way to securely share the dump with you?

I'll look into doing trying this tomorrow and also what I can do with the dump file I have. @KenKozman-GE any useful tips you have for loading this dll into the appservice from your attempts?

**Comment by KenKozman-GE** [comment]
Well we had to make our deployment `self-contained`, that was the biggest/gnarliest part, just because we were not already doing that. But it only took a few minutes to sort it out and have our pipeline allow for that. IIRC this was just a change to the `dotnet publish` command. Then we needed to include the clrgc.dll mentioned above.

The other part was setting the environment variable, but app settings also get set as environment variables, so that turned out to be easier than I thought it would be:

![image](https://user-images.githubusercontent.com/26578875/216174745-2a02239c-1fbe-4c55-89ea-4cc6d7780c68.png)

Let me know if I can help with any of this.

**Comment by Maoni0** [comment]
@Simon-Gregory-LG I'm currently seeking out a way to provide a secure link for you to share dumps. I should have an answer  tomorrow (looks like the person I need the info from already left for the day).

**Comment by Maoni0** [comment]
unfortunately my search for a standard way to let customers share files with us ended in vain. you could just use my google drive which I would need an email address to give you access to. feel free to send it to me to my work email (my email alias is maonis).

**Comment by Simon-Gregory-LG** [comment]
@Maoni0 many thanks for looking. I have messaged you independently regarding the memory dump.

I can also update you that I believe I've managed to deploy a version with the gc library running.

From Kudu process explorer:

![image](https://user-images.githubusercontent.com/45365503/216350849-725b02ee-2a95-4708-9caf-1a0c7eba5f76.png)

I'll update if I have any more information.

**Comment by KenKozman-GE** [comment]
TIL: You can find the modules in use via the Kudu process explorer! Thanks @Simon-Gregory-LG!

**Comment by cshung** [comment]
The earlier issue @Maoni0 discovered is fixed in https://github.com/dotnet/runtime/pull/81248.
This PR is currently under validation and we will merge it once the validation is over.

**Comment by Simon-Gregory-LG** [comment]
> TIL: You can find the modules in use via the Kudu process explorer! Thanks @Simon-Gregory-LG!

Haha, @KenKozman-GE thanks for your pointers by the way. Helped me get there quickly.

TIL: `'TIL'`!

**Comment by KenKozman-GE** [comment]
Well @Maoni0 it looks like I had the CPU peg issue happen again in our test instance, and the normal CPU monitoring was running so it captured a dump. Which I know is not what we were expecting to happen (vs. an intentional breakpoint/crash).

Also I can't seem to see much in this dump either (with managed debugging at least). But in native mode there are definitely a couple of threads in clrgc.dll.

So maybe this is the same bug as mentioned originally on the thread and it just did not trip the tripwires?

[Here](https://drive.google.com/file/d/14ehr_I-yQdFPmNdZrr8VXtBIVOCe14Ft/view?usp=sharing) is that dump. Let me know when you

**Comment by Maoni0** [comment]
@KenKozman-GE thanks! I was able to get the dump and just took a brief look. yes, exactly, this is the original issue and did not hit the instru I added which means it's likely somehow when the last GC ended it got a huge alloc context to begin with. I need to look at the dump some more and will report back.

**Comment by KenKozman-GE** [comment]
Awesome, we are getting closer. Well, and by we, I mean you guys :)

**Comment by Maoni0** [comment]
after some more investigation I don't think this has to do with alloc contexts because I don't think that first gen0 region was actually allocated between this and last GC, the problem is the gen0 region list does not make sense. at this point the plan_gen_num should still be 0 but it's already set to 1. in fact the gen0 regions (even though we just started processing gen0 regions)'s plan_gen_num's are all 1's. if we demoted them to gen0 the plan_gen_num should have been 0. so these should be in gen1 at the end of last GC. the last GC is always a gen2. so it looks like something went wrong during this gen2 GC but at this point I'm not sure how exactly it went wrong that caused it to go into this situation yet. 

I'll continue to investigate. I might produce another clrgc.dll if you wouldn't mind running with it. also disabling concurrent GC by setting `COMPlus_gcConcurrent` to 0 might help with reproing this faster.

**Comment by Simon-Gregory-LG** [comment]
I can also confirm the hang state occurred on ours yesterday on the instance with the clrgc.dll running and no exception was raised. 

I'll grab another memory dump and try that setting above.

**Comment by KenKozman-GE** [comment]
I didn’t understand 96% of what Maoni said but it sounds like it might be Gen2 GC related.

I will set that option as well to see if it repros more.

Happy to use a new/different GC DLL, just let us know.
________________________________
From: Simon-Gregory-LG ***@***.***>
Sent: Saturday, February 4, 2023 12:53:19 AM
To: dotnet/runtime ***@***.***>
Cc: Kozman, Ken (GE Digital) ***@***.***>; Mention ***@***.***>
Subject: EXT: Re: [dotnet/runtime] .NET 7 application hangs during GC (Issue #80073)

WARNING: This email originated from outside of GE. Please validate the sender's email address before clicking on links or attachments as they may not be safe.

I can also confirm the hang state occurred on ours yesterday on the instance with the clrgc.dll running and no exception was raised.

I'll grab another memory dump and try that setting above.

**Comment by Maoni0** [comment]
thank you both so much. @KenKozman-GE you are right - I don't need to see more dumps with the clrgc.dll you've been using. I suggested that 'cause I was hoping we could see a more frequent repro by disabling concurrent GC so when you use a new clrgc.dll we can shorten the time for the repro (it may not make any difference though because I'm seeing all your gen2s during the last 64 GCs are blocking and that part is also puzzling because the memory load is quite low. I will include the reasons why these are blocking in the new clrgc.dll).

I looked a bit more this morning and I'm pretty sure it was something went wrong during the last gen2 GC. the plan_gen_num problem I pointed out was actually not a problem (sorry, with a dump there's a lot of info that needs to be thought through in order to construct what has happened). so my current hypothesis is that during the last gen2 GC, which was a compacting and promoting GC, it somehow left that 1st region in gen0 which shouldn't have happened. I can see that region was all pinned during the last gen2 GC but our demotion logic (which can make a region go from a higher generation to a lower generation but that only happens if there's not much survival due to pinning) definitely shouldn't have demoted it. sorry my explanation may not make much sense without a lot of details (feel free to email me if you like to know any of these in more detail).

so to sum up, there are quite a few odd things in these dumps. I'll prepare a new clrgc.dll for both of you. ETA: tonight my time.

**Comment by KenKozman-GE** [comment]
No worries. I know I have baked my noodle on various post-mortem analyses and it can be sometimes fun and sometimes frustrating (I like to think that may be what being a detective is like). I've wasted a not insignificant portion of my life chasing down various threading bugs that have been gross (although perhaps not anywhere near as complex as the GC stuff). Nothing quite like the feeling when you finally get it all figured out though :)

No worries on the details, I can "sort of" follow the gist of it (and certainly as much as I need to). It all does sound quite interesting though :)

I'll try to get the clrgc.dll added in whenever you get it over to us, no rush.

**Comment by Maoni0** [comment]
I believe I know what the bug is. the problem is in `process_remaining_regions`, when we go through the remaining pinned plugs, for the region that contains the very last pinned plug, we just set its plan_gen_num to 0 (meaning to put this region in gen0, if we are doing a compacting GC). but this should have called `decide_on_demotion_pin_surv` instead (like the previous remaining regions with pinned plugs) which will decide whether we should put this region in gen0 based on how big pinned survival is. normally this would not cause a functional problem (it could cause a perf problem). but because there was that giant pinned plug it makes it not fit when it's unpinned and we call `allocate_in_condemned_generations` on it. so to make this bug show up you need the following conditions to all be true -

+ we did a gen2 compacting GC that generates a pinned plug in a gen2 region almost as big as the whole region. my guess for the reason why there's this giant pinned plug is because that gen2 region was already really compact so when we called `allocate_in_condemned_generations` on the non pinned plugs that are next to some pinned plugs in it we discovered we can't move the non pinned plugs anyway so we artificially pinned them and formed a giant pinned plug. and during this GC those objects were no longer pinned so we have one giant non pinned plug.
+ this gen2 region needs to be the last region with pinned plugs; 
+ this gen2 region hasn't been consumed by `allocate_in_condemned_generations` yet so it was processed by `process_remaining_regions`;

so that's very hard to hit. I'm making a fix. I need to test it before giving it to you so ETA tomorrow.

**Comment by Maoni0** [comment]
I debugged Simon's repro with him and he's seeing the exact same problem as what Ken is seeing. 

I have a new clrgc.dll at https://github.com/Maoni0/clrgc/tree/main/issues/80073/demotion_fix. this has the following 

+ a fix for the demotion problem I described above
+ a fix for the special sweep problem with Workstation GC
+ a bunch of instrumentation that helps to diagnose the perf problems I saw in dumps from Ken (why we did full blocking GCs, what created these pinned plugs, what was the situation when we turned on special sweep)
+ induce an AV when we see that a region is demoted to gen0 with large amount of pinned survival

if you had set the COMPlus_gcConcurrent env var to 0, please unset that. we don't need it.

if you could give this a whirl and let me know how it goes, that'd be much appreciated! thanks again.

**Comment by Simon-Gregory-LG** [comment]
Fantastic! I've got that deployed now:

![image](https://user-images.githubusercontent.com/45365503/217214368-08e50b6c-c25a-4551-b06c-4a34d6c5e6f5.png)

Have also removed the `COMPlus_gcConcurrent` environment variable.

So, I take it that with this fix deployed we should we be looking (and hoping :)) for the issue not to occur and for an access violation to not be raised?

**Comment by KenKozman-GE** [comment]
We are all patched with this as well (and I removed the `COMPlus_gcConcurrent` as well):

![image](https://user-images.githubusercontent.com/26578875/217315519-152b092d-632b-46c6-9b98-cee07c37235e.png)

I was wondering the same thing as @Simon-Gregory-LG as to what to expect now :)

**Comment by KenKozman-GE** [comment]
Okay @Maoni0 got a crash in `gc_heap::verify_regions` which is hopefully one of the instrumented bits that you are looking for.

I uploaded the dump to my Google drive and gave you access.

![image](https://user-images.githubusercontent.com/26578875/217336532-8ab6ec1f-c84f-4188-9ed7-5369ff44b78e.png)

(Side note: and I thought *I* had some large source files!)

**Comment by Maoni0** [comment]
oops, sorry I have been clear - you are not expected to see AV if the fix indeed addresses the problem. but if the fix didn't work, and it's due to the same bug, it would AV but would put us closer to failure so it's easier to see what happened. 

I just downloaded Ken's dump - this is again in verify_region and complaining that there are no regions in gen0 (same symptom  as what the 1st issue I discovered that's only specific to Workstation GC but not the same cause. I'll look more later this afternoon.

**Comment by Maoni0** [comment]
@KenKozman-GE your app is a good stress test for the GC 🙀 I think you are hitting an issue that we fixed in 8.0 - I believe the reason why it has 0 regions in gen0 is because we failed to get a new region for gen0. we are doing a sweeping gen1 GC in this case and there's plenty of space get a new region for gen0, but due to a bug we mistakenly didn't find enough space to commit the bookkeeping data for that region so we failed to get a new region -

```
0:047> dt clrgc!WKS::gc_heap::current_total_committed_bookkeeping 
0x1822000
0:047> dt clrgc!WKS::gc_heap::heap_hard_limit_for_bookkeeping
0x1824a2d
0:047> ? 0x1824a2d-0x1822000
Evaluate expression: 10797 = 00000000`00002a2d
```
that's not enough for a new region. let me make a new build with the fix for that. sorry about all the trouble!

**Comment by KenKozman-GE** [comment]
"Inadvertent stress test" sounds like a decent band name!

Just let me know when there is a fixed clrgc DLL and we can install it.

**Comment by Maoni0** [comment]
> "Inadvertent stress test" sounds like a decent band name!

I agree :)

I've ported 2 fixes from 8.0 in the latest clrgc.dll [#77480](https://github.com/dotnet/runtime/pull/77480) and [#80640](https://github.com/dotnet/runtime/pull/80640) (it's 2 fixes because in the 1st fix we got rid of one of 2 bookkeeping fields I mentioned above and that's relevant in the 2nd fix). 
please get the latest clrgc.dll from https://github.com/Maoni0/clrgc/tree/main/issues/80073/demotion_fix/v2. the file version should be (I increased the minor version number by 1)

`FileVersion     7,0,323,56101 @Commit: ac25991336fe96ed6892de68b28adc4756ab94a4`

thanks again for your patience!

**Comment by Simon-Gregory-LG** [comment]
> "Inadvertent stress test" sounds like a decent band name!
> 
> Just let me know when there is a fixed clrgc DLL and we can install it.

@KenKozman-GE  I also agree. I think there should be three albums: 'gen0', 'gen1' & 'gen2' then leave a gap to do a best of + unreleased materials called 'Unmanaged Heap'.

On a side note mine is still running fine with the first patch. Approaching 24 hours, but it's still within the window I've seen the issue occur. I might give it another 24 hours before I switch to try the new dll.

**Comment by KenKozman-GE** [comment]
Okay we are cooking with gas over here (have the latest clrgc.dll). Will wait and see what happens:

![image](https://user-images.githubusercontent.com/26578875/217598480-d438f204-ac4a-4faf-8e70-f68131fb31c3.png)

**Comment by rbouallou** [comment]
@Maoni0 Would it be possible to have a patched libclrgc.so for linux-x64?

**Comment by Maoni0** [comment]
@rbouallou, absolutely. I just put libclrgc.so along with src and symbols at https://github.com/Maoni0/clrgc/tree/main/issues/80073/demotion_fix/v2/linux. please let me know how this works out for you.

**Comment by Simon-Gregory-LG** [comment]
Just an update from me that the first patch (`7,0,323,56100`) has been running solidly for 3 days now. No issues so far (see below).

I will now install the newer (`7,0,323,56101`) to test over the weekend, but it's looking very good.

@Maoni0, thanks so much for the responsiveness and extremely helpful interactions on this issue so far, **it's been excellent**!

--- 
# Patch test summary:

## Health Check Metrics
* Red - .NET 7.0 version deployed
* Green - clrgc.dll patch `7,0,323,56100` deployed
> Note: Typically the app would become unhealthy within 36 hours

![image](https://user-images.githubusercontent.com/45365503/218078174-5ff28893-ace6-4277-b63b-ec9c8d957cdd.png)

## Crash Monitoring Summary

![image](https://user-images.githubusercontent.com/45365503/218078651-449b41eb-367f-44ad-8608-15a08c6bbfa8.png)

**Comment by Maoni0** [comment]
@Simon-Gregory-LG that's great to hear! and thank you so much for verifying :)

**Comment by Simon-Gregory-LG** [comment]
@Maoni0 no problem!

Lots going on today so forgot to post the confirmation actually, so here it is all up and running ready to enjoy the weekend!:

![image](https://user-images.githubusercontent.com/45365503/218187135-a6dc7965-981d-4074-b461-daa5c12dc04a.png)

**Comment by KenKozman-GE** [comment]
Wait... they give you guys weekends? I got to talk to my agent.

I just looked at our guinea pig instance over here and it is still rumbling along no crashes or infinite loops since the latest install. But I don't think our replication cycle was a reliable as @Simon-Gregory-LG. Maybe it will hit something this weekend.

**Comment by Simon-Gregory-LG** [comment]
So it's still going strong with 7,0,323,56101 and no unresponsiveness or exceptions (see below). Whereas our instances without the patch have gone down several times in this period.

We restarted it once to deploy the second patch, but we're again 3 days+ of stability with the new patch, so this looks very promising - amazing job @Maoni0 :)!

I take it that this hotfix is going to take a while before it makes the main .NET 7.0 runtimes and gets deployed to all the AppServices? In the interim should we consider this patch suitable for production deployment, or should we consider rolling back to .NET 6,0 for the time being?

---
# Health Check Metrics
Red - .NET 7.0 version deployed
Green - clrgc.dll patch 7,0,323,56100 deployed
Dark Green - clrgc.dll patch 7,0,323,56101 deployed

![image](https://user-images.githubusercontent.com/45365503/218469708-b0a35366-3447-411a-8256-373979bb2e80.png)

**Comment by KenKozman-GE** [comment]
@Simon-Gregory-LG: do you guys do self-contained deployments? If so I think the version of the framework is "baked in", so when/if this fix gets into .NET 7.0 we could just include that.

If not, we have seen where there is an app service extension that tends to be published (updated?) on/about when the latest version drop comes (e.g. 7.0.2 came out on Jan 10, 2023). Adding that extension has let us get the latest/greatest as the AppService included runtime seems to lag by a month or two (for testing and validation and what not I assume).

**Comment by InCerryGit** [comment]
> @rbouallou, absolutely. I just put libclrgc.so along with src and symbols at https://github.com/Maoni0/clrgc/tree/main/issues/80073/demotion_fix/v2/linux. please let me know how this works out for you.

Can this gc be used for .NET 6.0? we have .net6.0 projects using .net7.0 gc. @Maoni0

**Comment by Maoni0** [comment]
@InCerryGit would you mind opening a new issue for this as it has nothing with this issue? you could delete your comment after you open a new one (and I'll delete this comment too).

**Comment by InCerryGit** [comment]
> @InCerryGit would you mind opening a new issue for this as it has nothing with this issue? you could delete your comment after you open a new one (and I'll delete this comment too).
 
Sure，I'll delete it later. 
new issue: https://github.com/dotnet/runtime/issues/82416
Thanks. @Maoni0

**Comment by rbouallou** [comment]
> @rbouallou, absolutely. I just put libclrgc.so along with src and symbols at https://github.com/Maoni0/clrgc/tree/main/issues/80073/demotion_fix/v2/linux. please let me know how this works out for you.

We've been running with the patched libclrgc.so for just over a week now and it's still running with no issues - the process would hand every ~2 days with the default runtime.

@Maoni0 You mentioned fixes will be backported to 7.0. Do you have any idea when the release will come out?
Many thanks for the support.

**Comment by manandre** [comment]
> we will backport all fixes to 7.0 (the fix causing the infinite loop in `allocate_in_condemned_generations` that both Ken and Simon hit, and the other 2 showed up in `verify_regions` that only affected Ken).

@Maoni0 The next release for version 7.0 is planned on March 14th. Any chance these fixes will be part of it?

**Comment by Maoni0** [comment]
@manandre the reason why I haven't checked in is because I've been running more stress tests and hit multiple problems with that, - I hit some 8.0 problems (not in GC) when I tried to run stress on 8.0 so I ran this on 7.0 and I did hit another issue in the vicinity so I'd like to make a fix of that as well. should have it this week.

**Comment by Simon-Gregory-LG** [comment]
Hi @Maoni0, just wanted to quickly check if this fix has made its way into .NET 7 yet and if so, which version? 

I see this is referenced above in the PRs which were merged in April, but they look to be tagged with 7.0.7. So am I right in thinking that it's probably not quite yet, or did the fixes make it into the 7.0.5 one that's publicly available from April? (https://versionsof.net/core/7.0/)

**Comment by mangod9** [comment]
This fix should be available in 7.0.7 released last week: https://devblogs.microsoft.com/dotnet/june-2023-updates/

**Comment by Simon-Gregory-LG** [comment]
@mangod9 oh fantastic, hopefully that'll make it's way onto my Azure AppService soon. I'll keep an eye out for that.

Thanks for the quick reply! :)

**Comment by mangod9** [comment]
Closing since the fix has been made
